### ISSUE TITLE
Github Issues

### ISSUE THREAD

**Issue by andres-usheru** [healthchecks/healthchecks#671]
To create an issue in Github issues when an alert fires

**Comment by fukawi2** [comment]
This would be really useful for us too, so adding my +1

**Comment by JPaulMora** [comment]
I can help develop this, anything I should consider @cuu508?

**Comment by JPaulMora** [comment]
Okay I have started work on this, seems that the Github API does support what we're looking for, however I have some questions about implementation. 

Currently a single integration can be used for many checks, for that we'd have to consider

1. Either the PAT should have very broad permissions to address multimple issues across multiple user's repos
2. Setup each check with an optional github URL and PAT (this is the right technical solution, but how many checks would actually use this?)

**Comment by fukawi2** [comment]
That's basically what I was imagining. The "integration" would be configured to create the issue in a specific repository. Ideally with (optional) Labels added to the issue.

Commenting / closing it when the check goes back up would only be a "nice to have" feature IMO too.

**Comment by cuu508** [comment]
> I can help develop this, anything I should consider @cuu508?

Alright, happy to review a PR. 

There's a checklist for developing a new integration in [CONTRIBUTING.md](https://github.com/healthchecks/healthchecks/blob/master/CONTRIBUTING.md)

**Comment by cuu508** [comment]
Regarding labels, GitHub API docs have this note:

> Only users with push access can set labels for new issues. Labels are silently dropped otherwise.

Regarding commenting on issues, or closing issues when a check goes up, I think this will be tricky. When we create a new issue, GitHub will report back the issue number. We don't have a way to store the issue number, so we can find the right issue to comment on, or close later.

A sort-of similar integration is Trello. This integration only creates Trello cards, it does not comment on them or archive them.

**Comment by JPaulMora** [comment]
What I was thinking is, if we're gonna configure the repo and PAT at the notification level there would only be a single global repository we could notify issues to. eg I would not be able to monitor/alert multiple repos, right?

**Comment by cuu508** [comment]
@JPaulMora yes, but that's how the other integrations work as well: when you set up a Slack integration, you select a channel, and the integration later only posts to that channel. When you set up a Trello integration, you select a board and a list, and it only adds cards to that single list. When you set up a Telegram integration, you select a chat group, and the integration only posts to that one group.

Users can set up multiple integrations of the same type though, each with different settings. You could have two GitHub integrations in the same project, one creating issues in repo ABC, and the other creating issues in repo XYZ. Then you would assign one or the other integration to checks as needed.

**Comment by fukawi2** [comment]
> > Only users with push access can set labels for new issues. Labels are silently dropped otherwise.

Well at least if the PAT doesn't have the permissions, they would be silently dropped (rather than causing an error).

> What labels would we want to set? Isn't every repository different in how it's using the labels?

That would be configurable in the integration. In our specific case, we have various labels that are tied to webhooks that "do things". As an example, we have a label ("Infra") that when added to an issue, auto-assigns a random member of our Infra team.

So the workflow I'm envisaging with this integration is:
1. Check goes down
2. Create a GH Issue (with the appropriate labels, configured per integration)
3. Our auto-assigner randomly picks a team member to investigate the check.

We have 1 and 3, just need 2 to join them together 😄

**Comment by cuu508** [comment]
Revised flow:

* In Healthchecks, the user clicks on a "Connect GitHub" button
* Healthchecks performs an [OAuth web application flow](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-a-user-access-token-for-a-github-app#using-the-web-application-flow-to-generate-a-user-access-token) to get user's access token:
  * Healthchecks redirects the user to `https://github.com/login/oauth/authorize?client_id=...&state=....`
  * User accepts authorization request, GitHub redirects back to my app, and passes a code query parameter
  * Healthchecks exchanges the code for user access token ("ghu_***") by calling `https://github.com/login/oauth/access_token`
* Healthchecks looks up what installations the user has access to by calling `https://api.github.com/user/installations` and **using user's access token for authentication**
* For each installation, Healthchecks loads the list of repos by calling `https://api.github.com/user/installations/INSTALLATION_ID/repositories` and **using user's access token for authentication**
* Healthchecks presents the list of available repos to the user and asks the user to pick one
* The user picks a repo. Healthchecks stores the selected repo name and corresponding installation_id in the database

When Healthchecks presents the list of available repos, it should also include a comment along the lines of "Don't see the repo you are looking for? [Click Here] to install Healthchecks in you GithHub personal or organization account". The link would take the user to the GitHub app's installation page.

**Comment by fukawi2** [comment]
Amazing! I just tested it, it's perfect for us, thank-you! 👍 ❤
